core/src/test/java/io/confluent/kafka/schemaregistry/rest/json/RestApiTest.java

```diff
@@ -0,0 +1,232 @@
+      registerAndVerifySchema(restApp.restClient, schema, schemaIdCounter, subject2);
+      schemaIdCounter++;
+      allVersionsInSubject2.add(expectedVersion);
+    }
+    allSubjects.add(subject2);
+
+    // test getAllVersions with existing data
+    assertEquals(
+        "Getting all versions from subject1 should match all registered versions",
+        allVersionsInSubject1,
+        restApp.restClient.getAllVersions(subject1)
+    );
+    assertEquals(
+        "Getting all versions from subject2 should match all registered versions",
+        allVersionsInSubject2,
+        restApp.restClient.getAllVersions(subject2)
+    );
+
+    // test getAllSubjects with existing data
+    assertEquals("Getting all subjects should match all registered subjects",
+        allSubjects,
+        restApp.restClient.getAllSubjects()
+    );
+  }
+
+  @Test
+  public void testSchemaReferences() throws Exception {
+    Map<String, String> schemas = getJsonSchemaWithReferences();
+    String subject = "reference";
+    registerAndVerifySchema(restApp.restClient, schemas.get("ref.json"), 1, subject);
+
+    RegisterSchemaRequest request = new RegisterSchemaRequest();
+    request.setSchema(schemas.get("main.json"));
+    request.setSchemaType(JsonSchema.TYPE);
+    SchemaReference ref = new SchemaReference("ref.json", "reference", 1);
+    request.setReferences(Collections.singletonList(ref));
+    int registeredId = restApp.restClient.registerSchema(request, "referrer");
+    assertEquals("Registering a new schema should succeed", 2, registeredId);
+
+    SchemaString schemaString = restApp.restClient.getId(2);
+    // the newly registered schema should be immediately readable on the master
+    assertEquals("Registered schema should be found",
+        MAPPER.readTree(schemas.get("main.json")),
+        MAPPER.readTree(schemaString.getSchemaString())
+    );
+
+    assertEquals("Schema references should be found",
+        Collections.singletonList(ref),
+        schemaString.getReferences()
+    );
+  }
+
+  public static void registerAndVerifySchema(
+      RestService restService,
+      String schemaString,
+      int expectedId,
+      String subject
+  ) throws IOException, RestClientException {
+    int registeredId = restService.registerSchema(
+        schemaString,
+        JsonSchema.TYPE,
+        Collections.emptyList(),
+        subject
+    );
+    Assert.assertEquals(
+        "Registering a new schema should succeed",
+        (long) expectedId,
+        (long) registeredId
+    );
+    Assert.assertEquals("Registered schema should be found",
+        MAPPER.readTree(schemaString),
+        MAPPER.readTree(restService.getId(expectedId).getSchemaString())
+    );
+  }
+
+  public static List<String> getRandomJsonSchemas(int num) {
+    List<String> schemas = new ArrayList<>();
+    for (int i = 0; i < num; i++) {
+      String schema = "{\"type\":\"object\",\"properties\":{\"f"
+          + random.nextInt(Integer.MAX_VALUE)
```

Comment:
Nit: it's highly improbable, but we could end up with the same number twice here and break the tests every million builds or so :)